source/trace_event.c

```diff
@@ -0,0 +1,353 @@
+#include <errno.h>
+/* Disable warnings with windows build */
+#ifdef _MSC_VER
+#    pragma warning(disable : C4996)
+#    pragma warning(disable : C4100)
+#    pragma
+#endif
+
+/* use on external cJSON function for error checking */
+#define ERROR_CHECK(result, error_type)                                                                                \
+    do {                                                                                                               \
+        if (!result) {                                                                                                 \
+            aws_raise_error(error_type);                                                                               \
+            AWS_FATAL_ASSERT(0 && "trace failure");                                                                    \
+        }                                                                                                              \
+    } while (0)
+
+/*
+ * Data structures
+ */
+
+struct aws_trace_event {
+    struct cJSON *root, *event_array;
+    struct aws_bus bus;
+    struct aws_allocator *allocator;
+};
+
+struct aws_trace_event *trace_event;
+
+/*
+ * Place holder to be added in later
+
+    struct aws_trace_system_options{
+        bool write_to_file;
+        int time_units;
+    };
+ *
+ */
+// uint64_t listener_id = 1;
+uint64_t start_time;
+// bool write_to_file;
+char *filename_out;
+const int time_unit = AWS_TIMESTAMP_MICROS;
+
+struct aws_trace_event_metadata_args_1 {
+    int value;
+    char value_name[15];
+};
+struct aws_trace_event_metadata_args_2 {
+    int value[2];
+    char value_name[2][15];
+};
+
+struct aws_trace_event_metadata {
+    /* should be B/E for same scope or S/F for outside of scope */
+    char phase;
+    /* name of the event */
+    char name[15];
+    /* category of the event */
+    char category[15];
+    /* timestamp in milliseconds */
+    uint64_t timestamp;
+    uint64_t thread_id;
+    int process_id;
+    /* if args are enabled either args_1 or args_2 must be allocated */
+    bool args_enabled;
+
+    struct aws_trace_event_metadata_args_1 *args_1;
+
+    struct aws_trace_event_metadata_args_2 *args_2;
+};
+
+/*
+ * Private API
+ */
+
+/* function declaration to prevent compiler errors */
+void aws_trace_event_listener(uint64_t address, const void *msg, void *user_data);
+
+void aws_trace_system_write(void) {
+    if (time_unit == AWS_TIMESTAMP_NANOS) {
+        if (cJSON_AddStringToObject(trace_event->root, "displayTimeUnit", "ns") == NULL) {
+            aws_raise_error(AWS_ERROR_OOM);
+            AWS_FATAL_ASSERT(0 && "trace failure");
+        }
+    }
+    char *out = cJSON_Print((trace_event->root));
+    if (out == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+        AWS_FATAL_ASSERT(0 && "trace failure");
+    }
+    FILE *fp;
+    char fn[strlen(filename_out) + 6];
+
+    strcpy(fn, filename_out);
+    strncat(fn, ".json", 6);
```

Comment:
If you break out ".json" into a constant, you can grab the length from that and not have to hardcode the 6.